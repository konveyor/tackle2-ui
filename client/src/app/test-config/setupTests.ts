import "@testing-library/jest-dom";
import { server } from "@mocks/server";

const mockInitialized = false;

jest.mock("@react-keycloak/web", () => {
  const originalModule = jest.requireActual("@react-keycloak/web");
  return {
    ...originalModule,
    useKeycloak: () => [mockInitialized],
  };
});

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useLocation: () => ({
    pathname: "localhost:3000/example/path",
  }),
}));

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

// Uncomment the following to see request logging in the console:

// server.events.on("request:start", (req) => {
//   console.log(`Handling a request to ${req.url.href}`);
// });

// server.events.on("request:match", (req) => {
//   console.log(`Request to ${req.url.href} was matched with a handler`);
// });

// server.events.on("request:unhandled", (req) => {
//   console.warn(`Request to ${req.url.href} was not handled`);
// });
