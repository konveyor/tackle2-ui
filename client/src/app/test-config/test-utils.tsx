import React, { FC, ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { Application, Archetype, Assessment } from "@app/api/models";
import { RenderHookOptions, renderHook } from "@testing-library/react-hooks";

import { createContext, useContext } from "react";

const QueryClientContext = createContext<QueryClient | undefined>(undefined);

const AllTheProviders: FC<{
  children: React.ReactNode;
  queryClient?: QueryClient;
}> = ({ children, queryClient }) => {
  const internalQueryClient =
    queryClient ||
    new QueryClient({
      defaultOptions: {
        queries: {
          retry: false,
          cacheTime: 1000,
        },
      },
    });
  return (
    <QueryClientProvider client={internalQueryClient}>
      {children}
    </QueryClientProvider>
  );
};

export const useQueryClientContext = () => {
  const context = useContext(QueryClientContext);
  if (context === undefined) {
    throw new Error(
      "useQueryClientContext must be used within a QueryClientContext.Provider"
    );
  }
  return context;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

const customRenderHook = <TProps, TResult>(
  callback: (props: TProps) => TResult,
  options?: Omit<RenderHookOptions<TProps>, "wrapper"> & {
    queryClient?: QueryClient;
  }
) => {
  const { queryClient, ...rest } = options || {};
  const Wrapper: React.FC<{ children: React.ReactNode }> = ({ children }) => (
    <AllTheProviders queryClient={queryClient}>{children}</AllTheProviders>
  );

  return renderHook(callback, { wrapper: Wrapper as React.FC, ...rest });
};

export * from "@testing-library/react";

export { customRender as render };
export { customRenderHook as renderHook };

export const createMockAssessment = (
  overrides: Partial<Assessment> = {}
): Assessment => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    required: true,
    ...overrides,
  } as Assessment;
};

export const createMockApplication = (overrides: Partial<Application> = {}) => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    ...overrides,
  } as Application;
};

export const createMockArchetype = (overrides: Partial<Archetype> = {}) => {
  return {
    id: Math.random(),
    name: "Default name",
    description: "Default description",
    ...overrides,
  } as Archetype;
};
