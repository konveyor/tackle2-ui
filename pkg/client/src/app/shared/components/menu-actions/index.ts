export { MenuActions } from "./menu-actions";
