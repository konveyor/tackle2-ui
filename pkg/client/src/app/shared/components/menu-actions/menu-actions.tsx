import React, { useState } from "react";
import { Dropdown, DropdownItem, DropdownToggle } from "@patternfly/react-core";
import { CaretDownIcon } from "@patternfly/react-icons/dist/esm/icons/caret-down-icon";

export interface MenuActionsProps {
  actions: { label: string; callback: () => void }[];
}

export const MenuActions: React.FC<MenuActionsProps> = ({ actions }) => {
  const [isOpen, setIsOpen] = useState(false);

  return (
    <Dropdown
      isOpen={isOpen}
      onSelect={() => {
        setIsOpen(!isOpen);
      }}
      toggle={
        <DropdownToggle
          onToggle={(isOpen: boolean) => {
            setIsOpen(isOpen);
          }}
          toggleIndicator={CaretDownIcon}
        >
          Actions
        </DropdownToggle>
      }
      dropdownItems={actions.map((element, index) => (
        <DropdownItem key={index} component="button" onClick={element.callback}>
          {element.label}
        </DropdownItem>
      ))}
    />
  );
};
