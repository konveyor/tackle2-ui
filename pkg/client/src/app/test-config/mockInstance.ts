import MockAdapter from "axios-mock-adapter";
import axios from "axios";

export default new MockAdapter(axios);
