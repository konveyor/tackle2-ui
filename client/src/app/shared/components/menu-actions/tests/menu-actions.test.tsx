import React from "react";
import { MenuActions } from "../menu-actions";
import { render } from "@app/test-config/test-utils";

describe("MenuActions", () => {
  it("Renders without crashing", () => {
    const wrapper = render(
      <MenuActions
        actions={[
          { label: "Action1", callback: jest.fn() },
          { label: "Action2", callback: jest.fn() },
        ]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it("Executes callback", () => {
    const callback1Mock = jest.fn();
    const callback2Mock = jest.fn();

    render(
      <MenuActions
        actions={[
          { label: "Action1", callback: callback1Mock },
          { label: "Action2", callback: callback2Mock },
        ]}
      />
    );

    // Select dropdown btn
    // const dropdownBtn = wrapper.find("button").at(0);
    // expect(dropdownBtn.text()).toEqual("Actions");

    // // Verify callbacks are executed

    // dropdownBtn.simulate("click"); // Opens dropdown

    // const action1Btn = wrapper.find(".pf-c-dropdown__menu-item").at(0);
    // expect(action1Btn.text()).toEqual("Action1");
    // action1Btn.simulate("click");
    // expect(callback1Mock).toHaveBeenCalledTimes(1);

    // dropdownBtn.simulate("click"); // Opens dropdown
    // const action2Btn = wrapper.find(".pf-c-dropdown__menu-item").at(1);
    // expect(action2Btn.text()).toEqual("Action2");
    // action2Btn.simulate("click");
    // expect(callback2Mock).toHaveBeenCalledTimes(1);
  });
});
