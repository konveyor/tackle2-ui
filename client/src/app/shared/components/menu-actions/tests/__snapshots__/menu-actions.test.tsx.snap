// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenuActions Renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-c-dropdown"
        data-ouia-component-id="OUIA-Generated-Dropdown-1"
        data-ouia-component-type="PF4/Dropdown"
        data-ouia-safe="true"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="pf-c-dropdown__toggle"
          data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
          data-ouia-component-type="PF4/DropdownToggle"
          data-ouia-safe="true"
          id="pf-dropdown-toggle-id-0"
          type="button"
        >
          <span
            class="pf-c-dropdown__toggle-text"
          >
            Actions
          </span>
          <span
            class="pf-c-dropdown__toggle-icon"
          >
            <test-file-stub />
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-c-dropdown"
      data-ouia-component-id="OUIA-Generated-Dropdown-1"
      data-ouia-component-type="PF4/Dropdown"
      data-ouia-safe="true"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="pf-c-dropdown__toggle"
        data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
        data-ouia-component-type="PF4/DropdownToggle"
        data-ouia-safe="true"
        id="pf-dropdown-toggle-id-0"
        type="button"
      >
        <span
          class="pf-c-dropdown__toggle-text"
        >
          Actions
        </span>
        <span
          class="pf-c-dropdown__toggle-icon"
        >
          <test-file-stub />
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
